app/models/spree/payment.rb

```diff
@@ -141,7 +141,7 @@ def payment_source
 
     def ensure_correct_adjustment
       revoke_adjustment_eligibility if ['failed', 'invalid', 'void'].include?(state)
-      return if adjustment.try(:finalized?)
+      return if adjustment.try(:finalized?) || order.payment_state == 'credit_owed'
```

Comment:
This should work in most cases, but checking the state could go wrong. In theory, we could trigger another payment while credit is owed. Then we would owe even more. It doesn't seem to make sense at first but maybe we want to check if a payment goes through before adding an item.

Anyway, I think that the underlying assumption is that we don't pay transaction fees on refunds, no matter what the state of the order is. I would guess that we can check `amount <= 0`.

It also looks to me as if `revoke_adjustment_eligibility` would make sense in this case. So maybe it could look like this:
```rb
    def ensure_correct_adjustment
      revoke_adjustment_eligibility if failed_or_refund?
      ...
    end

    def failed_or_refund?
      ['failed', 'invalid', 'void'].include?(state) || amount <= 0
    end
```

Reply:
@mkllnk `revoke_adjustment_eligibility` looked like the obvious choice here, but it doesnt get executed when we use conditions like `amount <=0`, `order.payment_state == 'credit_owed` or `order.payment_state == paid`. These only work with the return keyword. I'm not exactly sure why 🤔

Reply:
I don't understand. The `revoke_adjustment_eligibility` should finalise the adjustment and the next line returns if the adjustment if finalised:
https://github.com/openfoodfoundation/openfoodnetwork/blob/1cbb0d8bbd8ee5d18405cf15c2e0992e336436e8/app/models/spree/payment.rb#L170-L178
https://github.com/openfoodfoundation/openfoodnetwork/blob/1cbb0d8bbd8ee5d18405cf15c2e0992e336436e8/app/models/spree/payment.rb#L143-L144

Can you write a spec first? Then we have an example to experiment with and you can show me what you mean.